### PR TITLE
CTC topology

### PR DESCRIPTION
**Pull request by eddy1021**
Add topology for Google CTC component

### CODE REVIEW

**Comment by lgirdwood** [comment]
@eddy1021 any updates or ETA ? do you think this will land in v2.10 ?

**Comment by cujomalainey** [comment]
> @eddy1021 any updates or ETA ? do you think this will land in v2.10 ?

We are blocked on internal build problems, I wouldnt expect it to be resolved for at least another 2 weeks.

File: tools/topology/topology2/cavs-mixin-mixout-ctc-ssp.conf
Comment (lgirdwood) [inline tools/topology/topology2/cavs-mixin-mixout-ctc-ssp.conf:40]: indentation looks a bit out.

File: tools/topology/topology2/cavs-mixin-mixout-ctc-ssp.conf
Comment (lgirdwood) [inline tools/topology/topology2/cavs-mixin-mixout-ctc-ssp.conf:5]: Do we need 2x gains ? I dont know the usage so I could be mistaken. Maybe best to explain usage in a comment above.

**Review by lgirdwood — COMMENTED** [review]
LGTM, but should get a review from @ranj063 and/or @singalsu

File: tools/topology/topology2/cavs-mixin-mixout-ctc-ssp.conf
Comment (eddy1021) [reply tools/topology/topology2/cavs-mixin-mixout-ctc-ssp.conf:5]: Thanks!. Add a comment.

File: tools/topology/topology2/cavs-mixin-mixout-ctc-ssp.conf
Comment (eddy1021) [reply tools/topology/topology2/cavs-mixin-mixout-ctc-ssp.conf:41]: Fixed it. Thanks for pointing this out.

File: tools/topology/topology2/include/components/ctc.conf
Comment (cujomalainey) [inline tools/topology/topology2/include/components/ctc.conf:16]: @lgirdwood @ranj063 is there a way to set a min/max channel in widgets? I don't see anything in widget common. Might be useful for validation when pipelines define channels outside the range their components support.

Comment (lgirdwood) [reply tools/topology/topology2/include/components/ctc.conf:16]: Adding @singalsu too

Comment (singalsu) [reply tools/topology/topology2/include/components/ctc.conf:16]: PCM has channels_min and channels_max but I don't know if it would here. In my experience when channels count can be other than default 2, it's better to list for the instances explicitly the in and out formats with possible channels counts (and all rates if non-48 kHz is possible). There's an arrays combiner than can help when there's many combinations.

Comment (cujomalainey) [reply tools/topology/topology2/include/components/ctc.conf:16]: Makes sense but is there a way to define that at the widget level.

E.g. dev A makes comp 1
Dev B makes pipeline with comp 1
Dev B fails to notice restrictions on formats

If format restrictions were part of the widget definition then it would be easy to identify if all widget formats were a superset of the pipelines set of supported formats. If they were a subset then the build would fail. Just a thought on finding bugs before runtime.

File: tools/topology/topology2/include/pipelines/cavs/mixout-gain-ctc-dai-copier-playback.conf
Comment (ranj063) [inline tools/topology/topology2/include/pipelines/cavs/mixout-gain-ctc-dai-copier-playback.conf:1]: @eddy1021 I have just introduced a new feature in the topology2 compiler to make it easier to extend extisting pipelines with one or more processing modules without having to copy & paste everything. For example, in your case you are adding the ctc module to the mixout-gain-dai-copier-playback.conf pipeline. Using the SubTreeCopy feature would make it easier for reviewers to see that you are simply extending the pipelines with an additional CTC module.

I have an example usage of this SubTreeCopy feature in my PR: https://github.com/thesofproject/sof/pull/9354. Please let me know if you have any questions.

File: tools/topology/topology2/include/components/ctc.conf
Comment (ranj063) [reply tools/topology/topology2/include/components/ctc.conf:16]: @cujomalainey audio formats are always part of individual widgets. A pipeline does not define a superset of audio formats at all. But its an interesting idea to identify discrepancies in audio formats across widgets within the same pipeline and fail during topology  build. Let me give it a thought.

File: tools/topology/topology2/include/pipelines/cavs/mixout-gain-ctc-dai-copier-playback.conf
Comment (eddy1021) [reply tools/topology/topology2/include/pipelines/cavs/mixout-gain-ctc-dai-copier-playback.conf:1]: Hi @ranj063 , I tried to use the SubTreeCopy feature([change](https://github.com/thesofproject/sof/pull/8840/commits/645b27acf7726b83f7865ade3a3f650681f5bafe)). But it will run into this error:
```
[171/267] Generating sof-mtl-rt1019-rt5682-ctc.tplg
ALSA lib pre-process-class.c:167:(tplg_class_get_unique_attribute_name) No unique attribute in class 'mixout-gain-ctc-dai-copier-playback'

ALSA lib pre-process-object.c:1844:(tplg_build_object) error setting unique attribute value for 'mixout-gain-ctc-dai-copier-playback.0'

ALSA lib pre-process-object.c:2083:(tplg_pre_process_objects) Error building object Pipeline.mixout-gain-ctc-dai-copier-playback.0

Unable to pre-process configuration
```

Not sure did I miss anything?

Comment (ranj063) [reply tools/topology/topology2/include/pipelines/cavs/mixout-gain-ctc-dai-copier-playback.conf:1]: @eddy1021 I just compiled your PR and it worked fine. Have you upgraded your alsa-utils to the latest upstream version? Alternatively, I think our CI docker image is also updated to be able to compile with the latest alsa-utils.

Comment (eddy1021) [reply tools/topology/topology2/include/pipelines/cavs/mixout-gain-ctc-dai-copier-playback.conf:1]: I updated `alsa-utils` and it's compiled now. Thanks!

**Review by ranj063 — APPROVED** [review]
Thanks for using the new feature @eddy1021. LGTM

**Comment by kv2019i** [comment]
@wszypelt this now pends on the docker/topology-toolchain update (like https://github.com/thesofproject/sof/pull/9354 ).

**Review by lgirdwood — APPROVED** [review]

**Review by cujomalainey — APPROVED** [review]

**Comment by cujomalainey** [comment]
according to #9354 toolchain has been updated, merging this now

**Pull request merged**
